File: solutions/Leetcode_109/leetcode_109.py
class Solution1:
    def sorted_listed_to_bst(self, head):
        if head:
            return self.bst(head, None)
        return None

    def bst(self, head, tail):
        if head == tail:
            return None
        fast = slow = head
        while fast != tail and fast.next != tail:
            fast = fast.next.next
            slow = slow.next

        node = TreeNode(slow.val)
        node.left = self.bst(head, slow)
        node.right = self.bst(slow.next, tail)
        return node

File: solutions/Leetcode_109/leetcode_109.cc
class Solution1
{
public:
    TreeNode *sortedListToBST(ListNode *head)
    {
        if (head == nullptr)
            return nullptr;
        return bst(head, nullptr);
    }

private:
    TreeNode *bst(ListNode *head, ListNode *tail)
    {
        if (head == tail)
            return nullptr;
        ListNode *fast = head, *slow = head;
        while (fast != tail && fast->next != tail)
        {
            fast = fast->next->next;
            slow = slow->next;
        }

        TreeNode *node = new TreeNode(slow->val);
        node->left = bst(head, slow);
        node->right = bst(slow->next, tail);
        return node;
    }
};